supervisor crashes if connection to server is closed

### Describe the bug
```
panic: send on closed channel

goroutine 28 [running]:
github.com/gitpod-io/gitpod/supervisor/pkg/gitpod.(*APIoverJSONRPC).handler(0xc000389ef0, 0xd9c720, 0xc000128000, 0xc00038a6c0, 0xc000640e10, 0x0, 0x0, 0x0, 0x0)
	/tmp/build/components-supervisor--app.40a74179369a0ebebb57101eda75e161b4fdb102/_deps/gitpod-core-components-supervisor--app/pkg/gitpod/gitpod-service.go:305 +0x2c5
github.com/sourcegraph/jsonrpc2.(*HandlerWithErrorConfigurer).Handle(0xc0004c8cc0, 0xd9c720, 0xc000128000, 0xc00038a6c0, 0xc000640e10)
	/workspace/go/pkg/mod/github.com/sourcegraph/jsonrpc2@v0.0.0-20200429184054-15c2290dcb37/handler_with_error.go:21 +0x73
github.com/sourcegraph/jsonrpc2.(*Conn).readMessages(0xc00038a6c0, 0xd9c720, 0xc000128000)
	/workspace/go/pkg/mod/github.com/sourcegraph/jsonrpc2@v0.0.0-20200429184054-15c2290dcb37/jsonrpc2.go:522 +0x57e
created by github.com/sourcegraph/jsonrpc2.NewConn
	/workspace/go/pkg/mod/github.com/sourcegraph/jsonrpc2@v0.0.0-20200429184054-15c2290dcb37/jsonrpc2.go:334 +0x1d5
{"@type":"type.googleapis.com/google.devtools.clouderrorreporting.v1beta1.ReportedErrorEvent","error":"exit status 2","message":"unexpected exit","ring":2,"serviceContext":{"service":"supervisor","version":""},"severity":"error","time":"2020-11-19T15:52:46Z"}
{"message":"sleeping five minutes to allow debugging","serviceContext":{"service":"supervisor","version":""},"severity":"info","time":"2020-11-19T15:52:46Z"}
```

--- a/components/supervisor/pkg/gitpod/gitpod-service.go
+++ b/components/supervisor/pkg/gitpod/gitpod-service.go
@@ -243,7 +243,7 @@ type APIoverJSONRPC struct {
 	C jsonrpc2.JSONRPC2
 
 	mu   sync.RWMutex
-	subs map[string][]chan *WorkspaceInstance
+	subs map[string]map[chan *WorkspaceInstance]struct{}
 }
 
 // Close closes the connection
@@ -262,24 +262,21 @@ func (gp *APIoverJSONRPC) InstanceUpdates(ctx context.Context, instanceID string
 
 	gp.mu.Lock()
 	if gp.subs == nil {
-		gp.subs = make(map[string][]chan *WorkspaceInstance)
+		gp.subs = make(map[string]map[chan *WorkspaceInstance]struct{})
 	}
-	gp.subs[instanceID] = append(gp.subs[instanceID], chn)
+	if sub, ok := gp.subs[instanceID]; ok {
+		sub[chn] = struct{}{}
+	} else {
+		gp.subs[instanceID] = map[chan *WorkspaceInstance]struct{}{chn: {}}
+	}
 	gp.mu.Unlock()
 
 	go func() {
 		<-ctx.Done()
 
 		gp.mu.Lock()
+		delete(gp.subs[instanceID], chn)
 		close(chn)
-		n := 0
-		for _, x := range gp.subs[instanceID] {
-			if x == chn {
-				continue
-			}
-			gp.subs[instanceID][n] = x
-			n++
-		}
 		gp.mu.Unlock()
 	}()
 
@@ -300,7 +297,7 @@ func (gp *APIoverJSONRPC) handler(ctx context.Context, conn *jsonrpc2.Conn, req 
 
 	gp.mu.RLock()
 	defer gp.mu.RUnlock()
-	for _, chn := range gp.subs[instance.ID] {
+	for chn := range gp.subs[instance.ID] {
 		select {
 		case chn <- &instance:
 		default:

